### PR TITLE
ADR for Automated Testing Strategy 

### PR DESCRIPTION
**Pull request by bgfoshay**
## What’s changing?

Create roadmap for implementing additional automated testing for the NAC.

## Why?

NAC-1188

## This PR has:

- [ ] **Commented source code** (_required on public classes and methods, and elsewhere as appropriate_)

- [x] **Developer documentation** (_for new build/test/API changes or complex portions of the system_)

- [ ] **Automated unit tests** (_to maintain or increase level of code coverage_)

- [ ] **Changes to IAC** (_add any deployment steps that require manual intervention to the draft release notes_)

### CODE REVIEW

**Review by adabbraccio-ventera — APPROVED** [review]
Thanks for letting me review! I agree with the approach.

Minor typo (dveloper instead of developer) in context section.

File: docs/adr/0035-testing-strategy.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+1. Write the minimum amount of code that fulfills the requirements of the ticket and satisfies the tests.
+1. Refactor and clean up the code.
+1. Repeat until all tests pass and the code is acceptable.
+
+Additionally, we strive to adhere to Martin Fowler’s definition of a “solitary” unit test for all of our unit tests. Our integration tests fall into the category of “socialable” tests that provide more end-to-end testing, and testing of downstream dependencies. In order to achieve “solitary” unit tests, we make heavy use of interfaces and abstract classes to decouple classes and functions. Not only is this a good design practice, but this approach creates a framework for us to make use of mocking libraries to create stubs, mocks, and test doubles. These are used to provide mock responses to API calls, API mocks to intercept and receive calls without invoking the actual API, and in-memory databases specifically for running with tests.
+
+For our testing, we use the following libraries
+
+* For our Blazor and javascript applications, we use bunit as our testing framework
```
Comment (dborgen2) [inline docs/adr/0035-testing-strategy.md:22]: Sidenote: The only thing we do with BUnit in regards to Javascript is making sure the Javascript function gets called from Blazor, as BUnit doesn't run in a browser so it doesn't test the Javascript code at all. However, that being said the Javascript code is very minor. I think we should do testing of the Javascript code with Cypress. Not sure if that affects the wording of this section at all, but wanted to point it out.

File: docs/adr/0035-testing-strategy.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+### Sept-Dec 2022
+* Currently we automatically build and run tests via CircleCI integration to GitHub. We will switch over to GitHub Enterprise and utilize GitHub Actions for running CICD pipelines.
+* Configure & document environment and run frequency for different categories of tests. e.g. Unit & integration tests will run against "Dev", "Test", and "Preprod" with every deployment, Performance tests will run against "Test" nightly, etc..
+* Establish connectivity between Test Runner(s) (in GitHub) and deployed applications so that network based tests (i.e. API tests) can be performed on the Test Runner(s)
+* Investigate approach to handle a test user and EAuth/PIV for end-to-end testing
+* Investigate approach to handle multiple tester user roles for end-to-end testing
+* Configure NetSparker to run Security Tests against our different environments
+* Begin incorporating API tests (Postman or RestAssured)
+* Create framework for running end-to-end tests that utilize a test user or users that have different roles
+* Incorporate more end-to-end and accessibility tests (cypress tests)
```
Comment (dborgen2) [inline docs/adr/0035-testing-strategy.md:58]: Do we want to note anything about mocking API endpoints with Cypress (assuming we switch to a SPA), or is that out of scope of this ADR?

**Comment by bgfoshay** [comment]
Placed a copy of the testing strategy at the root of the Engineering Team sharepoint folder.

**Pull request closed without merging**